pkg/config/virtio.go

```diff
@@ -181,6 +220,165 @@ func (dev *VirtioSerial) FromOptions(options []option) error {
+// to the machine.
+func VirtioInputNew(inputType string) (VirtioDevice, error) {
+	if inputType != VirtioInputPointingDevice && inputType != VirtioInputKeyboardDevice {
+		return nil, fmt.Errorf("Unknown option for virtio-input devices: %s", inputType)
+	}
+
+	return &VirtioInput{
+		InputType: inputType,
+	}, nil
+}
+
+func (dev *VirtioInput) validate() error {
+	if dev.InputType != VirtioInputPointingDevice && dev.InputType != VirtioInputKeyboardDevice {
+		return fmt.Errorf("Unknown option for virtio-input devices: %s", dev.InputType)
+	}
+
+	return nil
+}
+
+func (dev *VirtioInput) ToCmdLine() ([]string, error) {
+	if err := dev.validate(); err != nil {
+		return nil, err
+	}
+
+	return []string{"--device", fmt.Sprintf("virtio-input,%s", dev.InputType)}, nil
+}
+
+func (dev *VirtioInput) FromOptions(options []option) error {
+	for _, option := range options {
+		if option.key != VirtioInputPointingDevice && option.key != VirtioInputKeyboardDevice {
+			return fmt.Errorf("Unknown option for virtio-input devices: %s", option.key)
+		}
+		if option.value != "" {
+			return fmt.Errorf(fmt.Sprintf("Unexpected value for virtio-input %s option: %s", option.key, option.value))
+		}
+		dev.InputType = option.key
+	}
+	return dev.validate()
+}
+
+// VirtioGPUNew creates a new gpu device for the virtual machine.
+// The device parameter is the virtio-gpu device that will be added to the machine
+func VirtioGPUNew(device string) (VirtioDevice, error) {
+	if device != VirtioGPUDisplayDevice {
+		return nil, fmt.Errorf("Unknown option for virtio-gpu devices: %s", device)
+	}
+
+	virtioGPU := &VirtioGPU{
+		Device:  device,
+		UsesGUI: false,
+	}
+	virtioGPU.VirtioGPUResolution = VirtioGPUResolution{
+		Height: defaultVirtioGPUResolutionHeight,
+		Width:  defaultVirtioGPUResolutionWidth,
+	}
+	return virtioGPU, nil
+}
+
+// VirtioGPUNewWithGUI creates a new gpu device for the virtual machine with the
+// graphical user interface enabled. The device parameter is the virtio-gpu
+// device that will be added to the machine
+func VirtioGPUNewWithGUI(device string) (VirtioDevice, error) {
+	dev, err := VirtioGPUNew(device)
+	if err != nil {
+		return nil, err
+	}
+	dev.(*VirtioGPU).UsesGUI = true
+	return dev, nil
+}
+
+// VirtioGPUNewWithDimensions creates a new gpu device for the virtual machine
+// with user specified dimensions for the graphical user interface's resolution
+func VirtioGPUNewWithDimensions(device string, height, width int) (VirtioDevice, error) {
+	dev, err := VirtioGPUNew(device)
+	if err != nil {
+		return nil, err
+	}
+
+	dev.(*VirtioGPU).VirtioGPUResolution = VirtioGPUResolution{
+		Height: height,
+		Width:  width,
+	}
+	return dev, nil
+}
+
+// VirtioGPUNewWithGUIAndDimensions creates a new gpu device for the virtual
+// machine with the graphical user interface enabled and user defined
+// dimensions for its resolution.
+func VirtioGPUNewWithGUIAndDimensions(device string, height, width int) (VirtioDevice, error) {
```

Comment:
I don't think you need all these constructors as only one or 2 are used ? I'd prefer to have just VirtioGPUNew(...) with a set of args which are mandatory, and a few `VirtioGPUSetXxx` if that's needed.